ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/CommandRunner.java

```diff
@@ -83,19 +113,44 @@ public void close() {
   void fetchAndRunCommands() {
     final List<QueuedCommand> commands = commandStore.getNewCommands();
     log.trace("Found {} new writes to command topic", commands.size());
+    final AtomicBoolean shouldProcess = new AtomicBoolean(true);
     commands.stream()
         .filter(c -> c.getCommand().isPresent())
-        .forEach(c -> executeStatement(c.getCommand().get(), c.getCommandId(), c.getStatus()));
+        .forEach(c -> {
+          if (!shouldProcess.get()) {
+            return;
+          }
+          if (c.getCommand().get().getStatement()
+              .equalsIgnoreCase(TerminateCluster.TERMINATE_CLUSTER_STATEMENT_TEXT)) {
+            terminateCluster(c.getCommand().get());
+            shouldProcess.set(false);
+            return;
+          }
+          executeStatement(
+              c.getCommand().get(),
+              c.getCommandId(),
+              c.getStatus());
+        });
   }
 
   /**
    * Read and execute all commands on the command topic, starting at the earliest offset.
-   * @throws Exception TODO: Refine this.
    */
+  // TODO: creat github issue to handle terminate better.
   public void processPriorCommands() {
     final RestoreCommands restoreCommands = commandStore.getRestoreCommands();
+    final AtomicBoolean shouldProcess = new AtomicBoolean(true);
     restoreCommands.forEach(
         (commandId, command, terminatedQueries, wasDropped) -> {
```

Comment:
@big-andy-coates any suggestion on how to test this method? Didn't want to change it for easier testability before hearing from you.